vms/platformvm/state/state.go

```diff
@@ -943,74 +935,149 @@ func (s *state) ValidatorSet(subnetID ids.ID, vdrs validators.Set) error {
 	return nil
 }
 
-func (s *state) GetValidatorWeightDiffs(height uint64, subnetID ids.ID) (map[ids.NodeID]*ValidatorWeightDiff, error) {
-	prefixStruct := heightWithSubnet{
-		Height:   height,
-		SubnetID: subnetID,
-	}
-	prefixBytes, err := blocks.GenesisCodec.Marshal(blocks.Version, prefixStruct)
-	if err != nil {
-		return nil, err
-	}
-	prefixStr := string(prefixBytes)
-
-	if weightDiffs, ok := s.validatorWeightDiffsCache.Get(prefixStr); ok {
-		return weightDiffs, nil
-	}
-
-	rawDiffDB := prefixdb.New(prefixBytes, s.validatorWeightDiffsDB)
-	diffDB := linkeddb.NewDefault(rawDiffDB)
-	diffIter := diffDB.NewIterator()
+func (s *state) ApplyValidatorWeightDiffs(
+	validators map[ids.NodeID]*validators.GetValidatorOutput,
+	startHeight uint64,
+	endHeight uint64,
+	subnetID ids.ID,
+) error {
+	diffIter := s.flatValidatorWeightDiffsDB.NewIteratorWithStartAndPrefix(
+		getStartWeightKey(subnetID, startHeight),
+		subnetID[:],
+	)
 	defer diffIter.Release()
 
-	weightDiffs := make(map[ids.NodeID]*ValidatorWeightDiff)
+	prevHeight := startHeight + 1
 	for diffIter.Next() {
-		nodeID, err := ids.ToNodeID(diffIter.Key())
+		_, parsedHeight, nodeID, err := parseWeightKey(diffIter.Key())
 		if err != nil {
-			return nil, err
+			return err
 		}
+		if parsedHeight < endHeight {
```

Comment:
this is because heights are not contiguous right? We may have blocks with no stakers affected (not added, nor deleted) so we'd have a hole in the height list.